tests/test_testing.py

```diff
@@ -363,14 +363,55 @@ def test_httpclient_exception(pytester, book_list_html_response) -> None:
     }
     _save_fixture(
         pytester,
-        page_cls=ClientExceptionPage,
+        page_cls=ClientResponseErrorPage,
         page_inputs=[book_list_html_response, client],
         expected=item,
     )
     result = pytester.runpytest()
     result.assert_outcomes(passed=4)
 
 
+@attrs.define
+class ClientRequestErrorPage(WebPage):
+    client: HttpClient
+
+    async def to_item(self) -> dict:
+        msg = ""
+        try:
+            await self.client.get("http://books.toscrape.com/1.html")
+        except HttpRequestError as ex:
+            msg = ex.args[0]
+        return {"foo": "bar", "exception": msg}
+
+
+def test_httpclient_request_exception(pytester, book_list_html_response) -> None:
+    url = "http://books.toscrape.com/1.html"
+    request = HttpRequest(url)
+    exception = HttpRequestError("Bad Request", request)
+    responses = [
+        _SavedResponseData(request, None, exception),
+    ]
+    client = HttpClient(responses=responses)
+
+    item = {
+        "foo": "bar",
+        "exception": "Bad Request",
+    }
+    fixture = _save_fixture(
+        pytester,
+        page_cls=ClientRequestErrorPage,
+        page_inputs=[book_list_html_response, client],
+        expected=item,
+    )
+    assert (fixture.input_path / "HttpClient-0-exception_type.txt").exists()
+    assert (fixture.input_path / "HttpClient-0-exception.msg.txt").exists()
```

Comment:
Would it be possible to store the exception in a single file?

Reply:
It's possible, though we will need to stop using `serialize_leaf` for that and do the serialization inside `_serialize_HttpClient` (as to use the leaf serialization API we need to store the type name somewhere).

And I wonder what should we do with exceptions that have the `request` attribute non-empty. Currently it's serialized via the normal serialization API and so it's stored as separate file(s). Options:

1. Continue doing that, so that an exception will have its own data in a file created directly inside `_serialize_HttpClient` and also some file(s) for its embedded request.
2. Write some code inside `_serialize_HttpClient` that serializes the request into a single dict to put it inside the exception JSON.
3. Don't serialize `exception.request` and on deserialization set it to the request that this exception is related to (only if a flag that says "exception.request wasn't empty" is true?).

It's actually something worth discussing in any case: currently after deserialization `request` and `exception.request` will be different objects, I don't know if this can cause problems. Also note that scrapy-poet doesn't set the `request` attribute when raising `HttpError`.

Reply:
> It's possible, though we will need to stop using serialize_leaf for that and do the serialization inside _serialize_HttpClient

This looks ok to me, because exceptions are not leaf dependencies.

> And I wonder what should we do with exceptions that have the request attribute non-empty. Currently it's serialized via the normal serialization API and so it's stored as separate file(s).

A good catch. No strong opinion on that, but options 2 and 3 make more sense to me. (3) is probably fine.

> Also note that scrapy-poet doesn't set the request attribute when raising HttpError.

A good catch; looks like a bug?

Reply:
It may even be fine to not store any exception arguments now, because we're only checking the exception type.